tools/run_tests/xds_k8s_test_driver/framework/xds_k8s_testcase.py

```diff
@@ -301,55 +301,66 @@ def assertRpcStatusCodes(self,
                              method: str,
                              stray_rpc_limit: int = 0) -> None:
         """Assert all RPCs for a method are completing with a certain status."""
+        expected_status_int: int = expected_status.value[0]
+        expected_status_fmt: str = helpers_grpc.status_pretty(expected_status)
+
         # Sending with pre-set QPS for a period of time
         before_stats = test_client.get_load_balancer_accumulated_stats()
+        before_stats_fmt: str = self._pretty_accumulated_stats(before_stats)
         logging.debug(
```

Comment:
Respectfully disagree here. 

1. This was removed as a part of the effort of reducing the log size (per @larry-safran's feedback). I've spot-checked an Authz test suite log file ([authz_test_log_sample.log](https://github.com/grpc/grpc/files/11680373/authz_test_log_sample.log)), and these log statements take ~40% of by the number of lines, 1872 of 4652.
2. As you pointed out, this information is only relevant if we don't trust the diff function. 
3. I think the before/after logs may have a small relevance (f.e. to have a quick overview what other status codes the client encountered), but ONLY if we encountered an error. Therefore I log them as the last log message of the test, when the error happened, and clearly indicate this is for debugging purposes only.

Reply:
Thanks for the feedback.

> If we don't want to record them when they happen, then I don't think we'd log on error either. 

Just to clarify - we do log when the RPCs are sent, and when the after stats response received, just don't log the content:
```
I0607 02:35:43.005012 140519235196672 grpc.py:68] [psm-grpc-client-55d55774b4-d9mx9:8079] >> RPC LoadBalancerStatsService.GetClientAccumulatedStats(request=LoadBalancerAccumulatedStatsRequest({}), wait_for_ready=True, timeout=600)
I0607 02:35:43.511734 140519235196672 grpc.py:68] [psm-grpc-client-55d55774b4-d9mx9:8079] >> RPC LoadBalancerStatsService.GetClientAccumulatedStats(request=LoadBalancerAccumulatedStatsRequest({}), wait_for_ready=True, timeout=600)
I0607 02:35:43.518742 140519235196672 xds_k8s_testcase.py:320] [psm-grpc-client-55d55774b4-d9mx9] << Received accumulated stats difference. Expecting RPCs with status (0, OK) for method UNARY_CALL:
- method: UNARY_CALL
  rpcs_started: 13
  result:
    (0, OK): 13
```

> there's been no issue with the diffing and that is unrelated to these error messages

I agree 100% here, that was my reasoning why I decided agains logging before/after stats. The decision to log them at the end if an error happens was "just in case". 

> Either keep the logs, or trash them.

Sure, no problem, I'll remove the them.

> And certainly not logger.error() followed by raise with much of the content copied.

Just to clarify for future - is the problem in logger's level, or logger's level and the fact the most of the content is copied?

Reply:
I'd have had less of a problem with the log+raise if the log was info and just had the before/after details. I'd probably have turned my noise up at it a bit in the same "have it or don't have it, not based on the error" sense, but would have been much less likely to raise an objection. So it was both the logger's level and the fact much of the content was copied. And that's also partly the amount of code/formatting/duplication dedicated to the task, for the error case that had previously been just fine.

Reply:
Thanks. The reason I asked about the error log level and repeating the error is related to an idea I'm entertaining to address another point in @larry-safran's feedback - the one about not having error-level message when a test failed (which is certainly expected by folks familiar with absl-style logging). I was thinking to add the plumbing to automatically error-log the exception when a test just failed, before the teardown sequence. This also would solve several other problems as:

1. Having to determine that the test failed (or passed) from the fact the teardown sequence started, and not even seeing the error until all tests end.
2. The fact that python unittest (and absl) loggers just print (not log, so it doesn't even include the timestamp) errors collected from the all tests at the end of the log, after all tests has completed. This shows these errors out of context, and makes adds unnecessary extra steps for folks investigating an error. They to find the actual failed test among all other, find where it ends, scroll up above the teardown sequence, and find what was the last operation before the teardown. I have documented this process, but it's counterintuitive, and people tend to forget.
3. Because of (1) and (2) - not being able to clearly the timestamp of the error occurred. When I requested this info, I had to perform very complicated grep trickery to get the data, and then manually check if the error at the end corresponded to the failed test.

Reply:
I agree that not having the test failure in-line with the logs is unfortunate. Ordinarily for python testing there's not as much teardown and teardown logging as we do, so it makes sense to be a gap. If you think about JUnit, it doesn't include the test failure in-line either (although there's some Java issues that contribute to that as well).

I'd be quite fine with having the framework log an error. From a "ctrl-f to find the log" perspective, I don't know if it'd be best to duplicate the message or have a generic "THE TEST FAILED HERE", simply because there's two copies of the error on the sponge page already. But whatever, all that seems fine.

Reply:
Thanks! Btw, it's not just the teardown: when multiple tests executed, the error message will be separated from the test with the logs from other tests too (unless the failed test is the last one).
Example: https://source.cloud.google.com/results/invocations/a9e05410-b036-407a-a86f-b003a84ce92b/log, search for `[  FAILED  ]`.

The test failed at line 5400, the `[  FAILED  ]` appears after the tear down at line 5445, and the error + stack trace start as far as line 7025. So 1625 separate the place where the error happened, and where it's printed.

For this reason, I vote for duplicating the full stack trace (in our example it'd be line 5401). As a great side-effect, this will make the error more buggrep-friendly 😄

Reply:
SGTM